test-setup/src/client_setup.rs

```diff
@@ -0,0 +1,185 @@
+                        ClientFlow::receive_greeting(stream, self.client_flow_options)
+                            .await
+                            .unwrap();
+
+                    assert_eq!(expected_greeting, greeting);
+
+                    self.client = Client::Connected { flow };
+                }
+                Client::Connected { .. } => {
+                    panic!("Must be not connected for receiving greeting");
+                }
+                Client::Dummy => unreachable!(),
+            }
+        })
+    }
+
+    pub fn client_sends_command<'a>(&mut self, bytes: &'a [u8]) -> Command<'a> {
+        let command = self.decode_command(bytes);
+
+        self.rt.exec(async {
+            match &mut self.client {
+                Client::NotConnected { .. } => panic!("Must be connect for sending commands"),
+                Client::Connected { flow } => {
+                    let handle = flow.enqueue_command(command.to_static());
+                    let expected_event = ClientFlowEvent::CommandSent {
+                        handle,
+                        command: command.to_static(),
+                    };
+                    let event = flow.progress().await.unwrap();
+                    assert_eq!(expected_event, event);
+                }
+                Client::Dummy => unreachable!(),
+            }
+            self.server.stream.write_all(bytes).await.unwrap();
+        });
+
+        command
+    }
+
+    pub fn server_receives_command(&mut self, expected_command: Command) {
+        let bytes = self.command_codec.encode(&expected_command).dump();
+
+        self.rt.exec(self.server.read_expected_bytes(&bytes));
+    }
+}
+
+enum Client {
+    NotConnected { stream: TcpStream },
+    Connected { flow: ClientFlow },
+    Dummy,
+}
+
+impl Client {
+    async fn new(address: SocketAddr) -> Self {
+        let stream = TcpStream::connect(address).await.unwrap();
+        Self::NotConnected { stream }
+    }
+}
+
+struct Server {
+    stream: TcpStream,
+    read_buffer: BytesMut,
+}
+
+impl Server {
+    async fn new(listener: TcpListener) -> Self {
+        let (stream, _) = listener.accept().await.unwrap();
+        Self {
+            stream,
+            read_buffer: BytesMut::default(),
+        }
+    }
+
+    async fn read_expected_bytes(&mut self, expected_bytes: &[u8]) {
+        let expected_count = expected_bytes.len();
+
+        loop {
+            let bytes = &self.read_buffer[..];
+            let count = bytes.len();
+
+            if count < expected_count {
+                let expected_bytes = &expected_bytes[..count];
+
+                assert_eq!(expected_bytes, bytes);
+
+                self.stream.read_buf(&mut self.read_buffer).await.unwrap();
+            } else {
+                let bytes = &bytes[..expected_count];
+
+                assert_eq!(expected_bytes, bytes);
+
+                self.read_buffer.advance(expected_bytes.len());
+                break;
+            }
```

Comment:
question: Is the first `assert_eq!` necessary? From my understanding, this loop would repeatedly compare prefixes (but with a different length), i.e., 

```
aaaa[..1] != a
aaaa[..2] != aa
aaaa[..3] != aaa
aaaa[..4] != aaaa
```

But I'm tired, maybe I'm overlooking something :-)

Reply:
The whole purpose of this function is to assert the incoming bytes. The first `assert_eq!` is useful because it would abort earlier. If we expect banana but get only foo, then the second `assert_eq!` would never be reached.